server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,914 @@
+			result = transaction([data]);
+			break;
+		}
+	} catch (e) {
+		if (!e.name.endsWith('ErrorMessage')) {
+			Monitor.crashlog(e, 'A friends database process', query);
+			return {error: `Sorry! The database crashed. We've been notified and will fix this.`};
+		}
+		return {error: e.message};
+	}
+	return {...result} || {error: 'Unknown error in database query.'};
+});
+
+
+const ACTIONS = {
+	add: (
+		`REPLACE INTO friends (userid, friend, last_login) VALUES($userid, $friend, $login) ON CONFLICT (userid, friend) ` +
+		`DO UPDATE SET userid = $userid, friend = $friend`
+	),
+	get: `SELECT * FROM friends WHERE userid = ? LIMIT ?`,
+	delete: `DELETE FROM friends WHERE userid = $userid OR friend = $userid`,
+	getSent: `SELECT receiver, sender FROM friend_requests WHERE sender = ?`,
+	getReceived: `SELECT receiver, sender FROM friend_requests WHERE receiver = ?`,
+	insertRequest: `INSERT INTO friend_requests(sender, receiver, sent_at) VALUES(?, ?, ?)`,
+	deleteRequest: `DELETE FROM friend_requests WHERE sender = ? AND receiver = ?`,
+	findFriendship: `SELECT * FROM friends WHERE (friend = $user1 AND userid = $user2) OR (userid = $user1 AND friend = $user2)`,
+	findRequest: `SELECT * FROM friend_requests WHERE (sender = $user1 AND receiver = $user2) OR (sender = $user2 AND receiver = $user1)`,
+	renameFriend: `UPDATE OR IGNORE friends SET friend = $newID WHERE friend = $oldID`,
+	renameUserid: `UPDATE OR IGNORE friends SET userid = $newID WHERE userid = $oldID`,
+	rename: `REPLACE INTO friend_renames (original_name, new_name, change_date) VALUES(?, ?, ?)`,
+	login: `UPDATE friends SET last_login = ? WHERE friend = ?`,
+	checkLastLogin: `SELECT last_login FROM friends WHERE friend = ?`,
+	deleteLogin: `UPDATE friends SET last_login = null WHERE friend = ?`,
+};
+
+
+const TRANSACTIONS: {[k: string]: (input: any) => DatabaseResult} = {
+	rename: requests => {
+		for (const request of requests) {
+			const [oldID, newID] = request;
+			const oldFriends = statements.get.all(oldID, MAX_FRIENDS).map(item => item.friend);
+			const newRequests = statements.getReceived.all(oldID);
+			const sentRequests = statements.getSent.all(oldID);
+			if (newRequests.length || sentRequests.length) {
+				throw new Chat.ErrorMessage(`You have pending friend requests, and so cannot transfer friends.`);
+			}
+			if (!oldFriends.length) {
+				throw new Chat.ErrorMessage(`You have no friends to transfer.`);
+			}
+			if (oldFriends.includes(newID)) {
+				// unfriend the two
+				statements.delete.run({userid: newID});
+			}
+			statements.renameFriend.run({oldID, newID});
+			statements.renameUserid.run({oldID, newID});
+			statements.rename.run(oldID, newID, Date.now());
+		}
+		return {result: []};
+	},
+	add: requests => {
+		const addQuery = statements.add;
+		for (const request of requests) {
+			const [to, from] = request;
+			addQuery.run({userid: to, friend: from, login: Date.now()});
+			addQuery.run({userid: from, friend: to, login: Date.now()});
```

Comment:
It might not matter, but `Date.now()` isn't guaranteed the return the same value both times, I'd assign this to a variable once so all the rows have the same date